File: v1/src/main/java/com/api/v1/person/services/PersonModificationServiceImpl.java
package com.api.v1.person.services;

import com.api.v1.person.domain.Person;
import com.api.v1.person.domain.PersonAuditTrail;
import com.api.v1.person.domain.PersonAuditTrailRepository;
import com.api.v1.person.domain.PersonRepository;
import com.api.v1.person.dtos.PersonModificationRequestDto;
import jakarta.validation.Valid;
import jakarta.validation.constraints.NotNull;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Mono;

@Service
class PersonModificationServiceImpl implements PersonModificationService {

    @Autowired
    private PersonRepository personRepository;

    @Autowired
    private PersonAuditTrailRepository auditTrailRepository;

    @Override
    public Mono<Person> modify(@NotNull Person person, @Valid PersonModificationRequestDto requestDto) {
        return auditTrailRepository
                .save(PersonAuditTrail.of(person))
                .then(Mono.defer(() -> {
                    person.modify(requestDto);
                    return personRepository.save(person);
                }));
    }

}

File: kotlin/v3/src/main/kotlin/com/api/v3/persons/services/PersonModificationServiceImpl.kt
package com.api.v3.persons.services

import com.api.v3.persons.domain.Person
import com.api.v3.persons.domain.PersonRepository
import com.api.v3.persons.dtos.PersonModificationRequestDto
import jakarta.validation.Valid
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.jetbrains.annotations.NotNull
import org.springframework.stereotype.Service

@Service
internal class PersonModificationServiceImpl(
    private val personRepository: PersonRepository
): PersonModificationService {

    override suspend fun modify(
        @NotNull person: Person,
        requestDto: @Valid PersonModificationRequestDto
    ): Person {
        return withContext(Dispatchers.IO) {
            person.modify(requestDto)
            personRepository.save(person)
        }
    }

}